src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Globalization
+{
+    public partial class CultureInfo : IFormatProvider
+    {
+        internal static CultureInfo GetUserDefaultCulture()
+        {
+            if (GlobalizationMode.Invariant)
+                return CultureInfo.InvariantCulture;
+
+            string? strDefault = CultureData.GetLocaleInfoEx(Interop.Kernel32.LOCALE_NAME_USER_DEFAULT, Interop.Kernel32.LOCALE_SNAME);
+            if (strDefault == null)
+            {
+                strDefault = CultureData.GetLocaleInfoEx(Interop.Kernel32.LOCALE_NAME_SYSTEM_DEFAULT, Interop.Kernel32.LOCALE_SNAME);
+
+                if (strDefault == null)
+                {
+                    // If system default doesn't work, use invariant
+                    return CultureInfo.InvariantCulture;
+                }
+            }
+
+            return GetCultureByName(strDefault);
+        }
+
+        private static unsafe CultureInfo GetUserDefaultUICulture()
+        {
+            if (GlobalizationMode.Invariant)
+                return CultureInfo.InvariantCulture;
+
+            const uint MUI_LANGUAGE_NAME = 0x8;    // Use ISO language (culture) name convention
+            uint langCount = 0;
+            uint bufLen = 0;
+
+            if (Interop.Kernel32.GetUserPreferredUILanguages(MUI_LANGUAGE_NAME, &langCount, null, &bufLen) != Interop.BOOL.FALSE)
+            {
+                char[] languages = new char[bufLen];
```

Comment:
Would it be worth stackalloc'ing a span her for a short enough bufLen?  Or this is rare enough that it's not worth it?

Reply:
FWIW, I think that most of the time the buffer size here would be rather small, so you could indeed probably use a stack allocated buffer to save memory allocation most of the time. No idea how often this method is used though.

Reply:
GetUserDefaultUICulture expected to be called once as we cache it but I don't see a problem if we use stackalloc anyway 

---
In reply to: [418546679](https://github.com/dotnet/runtime/pull/35680#discussion_r418546679) [](ancestors = 418546679)